File: src/utils/getAllHandDiscardCombinations.ts
import { Card } from "../interfaces/card.interface";
import { getCombinations } from "./getCombinations";

// Helper function to check if two cards are the same (by rank and suit)
const areCardsEqual = (card1: Card, card2: Card): boolean => {
  return card1.rank === card2.rank && card1.suit === card2.suit;
};

export const getAllHandDiscardCombinations = (
  sixCards: Card[]
): { hand: Card[]; discard: Card[] }[] => {
  const allFourCardHands = getCombinations(sixCards, 4);
  const result: { hand: Card[]; discard: Card[] }[] = [];

  for (const fourCardHand of allFourCardHands) {
    const discard: Card[] = sixCards.filter(
      (card) => !fourCardHand.some((handCard) => areCardsEqual(card, handCard))
    );
    result.push({ hand: fourCardHand, discard: discard });
  }

  return result;
};

File: crib_calc/utils/get_all_hand_discard_combinations.py
from crib_calc.card import Card
from crib_calc.utils.get_combinations import get_combinations


def get_all_hand_discard_combinations(
    six_cards: list[Card],
) -> list[dict[str, list[Card]]]:
    result = []
    for hand in get_combinations(six_cards, 4):
        hand_set = {(c.rank, c.suit) for c in hand}
        discard = [c for c in six_cards if (c.rank, c.suit) not in hand_set]
        result.append({"hand": hand, "discard": discard})
    return result
